File: src/main/java/de/tectoast/emolga/utils/RequestBuilder.java
package de.tectoast.emolga.utils;


import com.google.api.services.sheets.v4.Sheets;
import com.google.api.services.sheets.v4.model.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

import static de.tectoast.emolga.commands.Command.*;
import static de.tectoast.emolga.utils.Google.getSheetsService;


@SuppressWarnings("UnusedReturnValue")
public class RequestBuilder {

    private static final Logger logger = LoggerFactory.getLogger(RequestBuilder.class);

    private final ArrayList<MyRequest> requests = new ArrayList<>();
    private final String sid;
    private boolean executed = false;
    private Runnable runnable;
    private long delay = 0;
    private boolean suppressMessages = false;
    private String[] additionalSheets;

    /**
     * Creates a RequestBuilder
     *
     * @param sid The ID of the sheet where the values should be written
     */
    public RequestBuilder(String sid) {
        this.sid = sid;
    }

    public static void updateSingle(String sid, String range, Object value, boolean... raw) {
        updateRow(sid, range, Collections.singletonList(value), raw);
    }

    public static void updateRow(String sid, String range, List<Object> values, boolean... raw) {
        updateAll(sid, range, Collections.singletonList(values), raw);
    }

    public static void updateAll(String sid, String range, List<List<Object>> values, boolean... raw) {
        new Thread(() -> {
            try {
                getSheetsService().spreadsheets().values().update(sid, range, new ValueRange().setValues(values)).setValueInputOption(raw.length == 0 || !raw[0] ? "USER_ENTERED" : "RAW").execute();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }, "ReqBuilder").start();
    }

    public static void batchUpdate(String sid, Request... requests) {
        if (requests.length == 0) return;
        try {
            getSheetsService().spreadsheets().batchUpdate(sid, new BatchUpdateSpreadsheetRequest()
                    .setRequests(Arrays.asList(requests))).execute();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static int getColumnFromRange(String range) {
        char[] chars = range.replaceAll("[^a-zA-Z]", "").toCharArray();
        if (chars.length == 1) return chars[0] - 65;
        return (chars[0] - 64) * 26 + (chars[1] - 65);
    }

    public static int getRowFromRange(String range) {
        return Integer.parseInt(range.replaceAll("[^\\d]", "")) - 1;
    }

    public static GridRange buildGridRange(String expr, int sheetId) {
        String[] split = expr.split(":");
        String s1 = split[0];
        String s2 = split.length == 1 ? s1 : split[1];
        GridRange r = new GridRange();
        r.setSheetId(sheetId);
        r.setStartColumnIndex(getColumnFromRange(s1))
                .setStartRowIndex(getRowFromRange(s1));
        r.setEndColumnIndex(getColumnFromRange(s2) + 1)
                .setEndRowIndex(getRowFromRange(s2) + 1);
        try {
            logger.info("r = {}", r.toPrettyString());
        } catch (IOException e) {
            e.printStackTrace();
        }
        return r;
    }

    public RequestBuilder withRunnable(Runnable r) {
        this.runnable = r;
        return this;
    }

    public RequestBuilder withRunnable(Runnable r, long delay) {
        this.runnable = r;
        this.delay = delay;
        return this;
    }

    public RequestBuilder suppressMessages() {
        this.suppressMessages = true;
        return this;
    }

    public RequestBuilder withAdditionalSheets(String... additionalSheets) {
        this.additionalSheets = additionalSheets;
        return this;
    }

    /**
     * Adds a single object to the builder
     *
     * @param range The range, where the object should be written
     * @param body  The single object that should be written
     * @param raw   optional argument, which makes the request raw (if true) or user entered (if false or null)
     * @return this RequestBuilder
     */
    public RequestBuilder addSingle(String range, Object body, boolean... raw) {
        return addRow(range, Collections.singletonList(body), raw);
    }

    /**
     * Adds a row of objects to the builder
     *
     * @param range The range, where the row should be written
     * @param body  The row that should be written
     * @param raw   optional argument, which makes the request raw (if true) or user entered (if false or null)
     * @return this RequestBuilder
     */
    @SuppressWarnings("UnusedReturnValue")
    public RequestBuilder addRow(String range, List<Object> body, boolean... raw) {
        return addAll(range, Collections.singletonList(body), raw);
    }

    /**
     * Adds a matrix of objects to the builder
     *
     * @param range The range, where the objects should be written
     * @param body  The matrix that should be written
     * @param raw   optional argument, which makes the request raw (if true) or user entered (if false or null)
     * @return this RequestBuilder
     */
    public RequestBuilder addAll(String range, List<List<Object>> body, boolean... raw) {
        requests.add(new MyRequest().setRange(range).setSend(body).setValueInputOption(raw.length == 0 || !raw[0] ? ValueInputOption.USER_ENTERED : ValueInputOption.RAW));
        return this;
    }

    public RequestBuilder addColumn(String range, List<Object> body, boolean... raw) {
        return addAll(range, body.stream().map(Collections::singletonList).collect(Collectors.toList()));
    }

    /**
     * Adds one or multiple batch request(s) to the builder
     *
     * @param requests The request(s) that should be sent
     * @return this RequestBuilder
     */
    public RequestBuilder addBatch(Request... requests) {
        Arrays.stream(requests).map(r -> new MyRequest().setRequest(r)).forEach(this.requests::add);
        return this;
    }

    public RequestBuilder addBGColorChange(int sheetId, String range, Color c) {
        String[] split = range.split(":");
        String s1 = split[0];
        String s2 = split.length == 1 ? s1 : split[1];
        return addBatch(new Request().setUpdateCells(new UpdateCellsRequest()
                .setRange(buildGridRange(range, sheetId))
                .setFields("userEnteredFormat.backgroundColor")
                .setRows(getCellsAsRowData(
                        new CellData().setUserEnteredFormat(new CellFormat().setBackgroundColor(c)), getColumnFromRange(s2) - getColumnFromRange(s1) + 1, getRowFromRange(s2) - getRowFromRange(s1) + 1
                ))));
    }

    public RequestBuilder addBGColorChange(int sheetId, int x, int y, Color c) {
        return addBGColorChange(sheetId, getAsXCoord(x) + y, c);
    }

    public RequestBuilder addNoteChange(int sheetId, String range, String note) {
        String[] split = range.split(":");
        String s1 = split[0];
        String s2 = split.length == 1 ? s1 : split[1];
        return addBatch(new Request().setUpdateCells(new UpdateCellsRequest()
                .setRange(buildGridRange(range, sheetId))
                .setFields("note")
                .setRows(getCellsAsRowData(
                        new CellData().setNote(note), getColumnFromRange(s2) - getColumnFromRange(s1) + 1, getRowFromRange(s2) - getRowFromRange(s1) + 1
                ))));
    }

    public RequestBuilder addHorizontalAlignmentChange(int sheetId, String range, String alignment) {
        String[] split = range.split(":");
        String s1 = split[0];
        String s2 = split.length == 1 ? s1 : split[1];
        return addBatch(new Request().setUpdateCells(new UpdateCellsRequest()
                .setRange(buildGridRange(range, sheetId))
                .setFields("userEnteredFormat.horizontalAlignment")
                .setRows(getCellsAsRowData(
                        new CellData().setUserEnteredFormat(new CellFormat().setHorizontalAlignment(alignment)), getColumnFromRange(s2) - getColumnFromRange(s1) + 1, getRowFromRange(s2) - getRowFromRange(s1) + 1
                ))));
    }

    public RequestBuilder addVerticalAlignmentChange(int sheetId, String range, String alignment) {
        String[] split = range.split(":");
        String s1 = split[0];
        String s2 = split.length == 1 ? s1 : split[1];
        return addBatch(new Request().setUpdateCells(new UpdateCellsRequest()
                .setRange(buildGridRange(range, sheetId))
                .setFields("userEnteredFormat.verticalAlignment")
                .setRows(getCellsAsRowData(
                        new CellData().setUserEnteredFormat(new CellFormat().setVerticalAlignment(alignment)), getColumnFromRange(s2) - getColumnFromRange(s1) + 1, getRowFromRange(s2) - getRowFromRange(s1) + 1
                ))));
    }

    public RequestBuilder addFontChange(int sheetId, String range, String font) {
        String[] split = range.split(":");
        String s1 = split[0];
        String s2 = split.length == 1 ? s1 : split[1];
        return addBatch(new Request().setUpdateCells(new UpdateCellsRequest()
                .setRange(buildGridRange(range, sheetId))
                .setFields("userEnteredFormat.textFormat.fontFamily")
                .setRows(getCellsAsRowData(
                        new CellData().setUserEnteredFormat(new CellFormat().setTextFormat(new TextFormat().setFontFamily(font))), getColumnFromRange(s2) - getColumnFromRange(s1) + 1, getRowFromRange(s2) - getRowFromRange(s1) + 1
                ))));
    }

    public RequestBuilder addStrikethroughChange(int sheetId, String range, boolean strikethrough) {
        String[] split = range.split(":");
        String s1 = split[0];
        String s2 = split.length == 1 ? s1 : split[1];
        return addBatch(new Request().setUpdateCells(new UpdateCellsRequest()
                .setRange(buildGridRange(range, sheetId))
                .setFields("userEnteredFormat.textFormat.strikethrough")
                .setRows(getCellsAsRowData(
                        new CellData().setUserEnteredFormat(new CellFormat().setTextFormat(new TextFormat().setStrikethrough(strikethrough))), getColumnFromRange(s2) - getColumnFromRange(s1) + 1, getRowFromRange(s2) - getRowFromRange(s1) + 1
                ))));
    }

    public RequestBuilder addStrikethroughChange(int sheetId, int x, int y, boolean strikethrough) {
        return addStrikethroughChange(sheetId, getAsXCoord(x) + y, strikethrough);
    }

    public RequestBuilder addFGColorChange(int sheetId, String range, Color c) {
        String[] split = range.split(":");
        String s1 = split[0];
        String s2 = split.length == 1 ? s1 : split[1];
        return addBatch(new Request().setUpdateCells(new UpdateCellsRequest()
                .setRange(buildGridRange(range, sheetId))
                .setFields("userEnteredFormat.textFormat.foregroundColor")
                .setRows(getCellsAsRowData(
                        new CellData().setUserEnteredFormat(new CellFormat().setTextFormat(new TextFormat().setForegroundColor(c))), getColumnFromRange(s2) - getColumnFromRange(s1) + 1, getRowFromRange(s2) - getRowFromRange(s1) + 1
                ))));
    }

    public RequestBuilder addFGColorChange(int sheetId, int x, int y, Color c) {
        return addFGColorChange(sheetId, getAsXCoord(x) + y, c);
    }


    private List<ValueRange> getUserEntered() {
        return requests.stream().filter(r -> r.valueInputOption.equals(ValueInputOption.USER_ENTERED)).map(MyRequest::build).collect(Collectors.toList());
    }

    private List<ValueRange> getRaw() {
        return requests.stream().filter(r -> r.valueInputOption.equals(ValueInputOption.RAW)).map(MyRequest::build).collect(Collectors.toList());
    }

    private List<Request> getBatch() {
        return requests.stream().filter(r -> r.valueInputOption.equals(ValueInputOption.BATCH)).map(MyRequest::buildBatch).collect(Collectors.toList());
    }

    /**
     * Executes the request to the specified google sheet
     */
    public void execute() {
        if (executed)
            throw new IllegalStateException("Already executed RequestBuilder with requests:\nsid = " + this.sid + "\n" + requests);

        List<ValueRange> userentered = getUserEntered();
        List<ValueRange> raw = getRaw();
        List<Request> batch = getBatch();
        Sheets service = getSheetsService();
        List<Thread> list = new LinkedList<>();

        list.add(new Thread(() -> {
            if (!userentered.isEmpty()) {
                if (!suppressMessages)
                    for (int i = 0; i < userentered.size(); i++) {
                        ValueRange range = userentered.get(i);
                        logger.info("{}: {} -> {}", i, range.getRange(), range.getValues());
                    }
                try {
                    service.spreadsheets().values().batchUpdate(sid, new BatchUpdateValuesRequest().setData(userentered).setValueInputOption("USER_ENTERED")).execute();
                    if (additionalSheets != null) {
                        for (String sidd : additionalSheets) {
                            service.spreadsheets().values().batchUpdate(sidd, new BatchUpdateValuesRequest().setData(userentered).setValueInputOption("USER_ENTERED")).execute();
                        }
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                    sendStacktraceToMe(e);
                }
            }
        }, "ReqBuilder User"));
        list.add(new Thread(() -> {
            if (!raw.isEmpty()) {
                try {
                    service.spreadsheets().values().batchUpdate(sid, new BatchUpdateValuesRequest().setData(raw).setValueInputOption("RAW")).execute();
                    if (additionalSheets != null) {
                        for (String sidd : additionalSheets) {
                            service.spreadsheets().values().batchUpdate(sidd, new BatchUpdateValuesRequest().setData(raw).setValueInputOption("RAW")).execute();
                        }
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                    sendStacktraceToMe(e);
                }
            }
        }, "ReqBuilder Raw"));
        list.add(new Thread(() -> {
            if (!batch.isEmpty()) {
                try {
                    service.spreadsheets().batchUpdate(sid, new BatchUpdateSpreadsheetRequest().setRequests(batch)).execute();
                    if (additionalSheets != null) {
                        for (String sidd : additionalSheets) {
                            service.spreadsheets().batchUpdate(sidd, new BatchUpdateSpreadsheetRequest().setRequests(batch)).execute();
                        }
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                    sendStacktraceToMe(e);
                }
            }
        }, "ReqBuilder Batch"));
        list.forEach(Thread::start);
        //noinspection IfStatementWithIdenticalBranches
        if (runnable != null) {
            try {
                for (Thread thread : list) {
                    thread.join();
                }
                Thread.sleep(delay);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            runnable.run();
            executed = true;
        } else
            executed = true;
    }

    private enum ValueInputOption {
        RAW,
        USER_ENTERED,
        BATCH
    }

    private static class MyRequest {
        private String range;
        private List<List<Object>> send;
        private ValueInputOption valueInputOption;
        private Request request;

        public MyRequest setRange(String range) {
            this.range = range;
            return this;
        }

        public MyRequest setSend(List<List<Object>> send) {
            this.send = send;
            return this;
        }

        public MyRequest setRequest(Request request) {
            this.request = request;
            this.valueInputOption = ValueInputOption.BATCH;
            return this;
        }

        public MyRequest setValueInputOption(ValueInputOption valueInputOption) {
            this.valueInputOption = valueInputOption;
            return this;
        }

        public ValueRange build() {
            return new ValueRange().setValues(send).setRange(range);
        }

        public Request buildBatch() {
            return request;
        }

        @Override
        public String toString() {
            return "Request{" +
                    "range='" + range + '\'' +
                    ", send=" + send +
                    ", valueInputOption='" + valueInputOption + '\'' +
                    '}';
        }
    }
}

File: src/main/kotlin/de/tectoast/emolga/utils/RequestBuilder.kt
package de.tectoast.emolga.utils

import com.google.api.client.googleapis.json.GoogleJsonResponseException
import com.google.api.services.sheets.v4.model.*
import de.tectoast.emolga.utils.records.Coord
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.async
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.slf4j.LoggerFactory
import java.util.regex.Pattern

@Suppress("unused")
class RequestBuilder
/**
 * Creates a RequestBuilder
 *
 * @param sid The ID of the sheet where the values should be written
 */(val sid: String) {
    private val requests: MutableList<MyRequest> = ArrayList()
    private var executed = false
    private var runnable: (suspend () -> Unit)? = null
    private var delay: Long = 0
    private var suppressMessages = false
    private var additionalSheets: Array<String>? = null
    private var onlyBatch = false
    private var realExecute = true
    fun withRunnable(r: suspend () -> Unit): RequestBuilder {
        runnable = r
        return this
    }

    fun withRunnable(delay: Long, r: suspend () -> Unit): RequestBuilder {
        runnable = r
        this.delay = delay
        return this
    }

    fun suppressMessages(): RequestBuilder {
        suppressMessages = true
        return this
    }

    fun withAdditionalSheets(vararg additionalSheets: String): RequestBuilder {
        this.additionalSheets = arrayOf(*additionalSheets)
        return this
    }

    fun onlyBatch(): RequestBuilder {
        onlyBatch = true
        return this
    }

    /**
     * Adds a single object to the builder
     *
     * @param range The range, where the object should be written
     * @param body  The single object that should be written
     * @param raw   optional argument, which makes the request raw (if true) or user entered (if false or null)
     * @return this RequestBuilder
     */
    fun addSingle(range: String?, body: Any, raw: Boolean = false): RequestBuilder {
        return addRow(range, listOf(body), raw)
    }

    fun addSingle(range: Coord, body: Any, raw: Boolean = false): RequestBuilder {
        return addSingle(range.toString(), body, raw)
    }

    /**
     * Adds a row of objects to the builder
     *
     * @param range The range, where the row should be written
     * @param body  The row that should be written
     * @param raw   optional argument, which makes the request raw (if true) or user entered (if false or null)
     * @return this RequestBuilder
     */
    fun addRow(range: String?, body: List<Any>, raw: Boolean = false): RequestBuilder {
        return addAll(range, listOf(body), raw)
    }

    fun addRow(range: Coord, body: List<Any>, raw: Boolean = false): RequestBuilder {
        return addRow(range.toString(), body, raw)
    }

    /**
     * Adds a matrix of objects to the builder
     *
     * @param range The range, where the objects should be written
     * @param body  The matrix that should be written
     * @param raw   optional argument, which makes the request raw (if true) or user entered (if false or null)
     * @return this RequestBuilder
     */
    fun addAll(range: String?, body: List<List<Any>?>?, raw: Boolean = false): RequestBuilder {
        requests.add(
            MyRequest().setRange(range).setSend(body)
                .setValueInputOption(if (raw) ValueInputOption.RAW else ValueInputOption.USER_ENTERED)
        )
        return this
    }

    fun addAll(range: Coord, body: List<List<Any>?>?, raw: Boolean = false): RequestBuilder {
        return addAll(range.toString(), body, raw)
    }

    fun addColumn(range: String?, body: List<Any>, raw: Boolean = false): RequestBuilder {
        return addAll(
            range, body.map { listOf(it) }, raw
        )
    }

    fun addColumn(range: Coord, body: List<Any>, raw: Boolean = false): RequestBuilder {
        return addColumn(range.toString(), body, raw)
    }

    /**
     * Adds one or multiple batch request(s) to the builder
     *
     * @param requests The request(s) that should be sent
     * @return this RequestBuilder
     */
    fun addBatch(vararg requests: Request): RequestBuilder {
        requests.map { MyRequest().setRequest(it) }.forEach { this.requests.add(it) }
        return this
    }

    private fun getCellsAsRowData(cellData: CellData, x: Int, y: Int) =
        List(y) { RowData().setValues(List(x) { cellData }) }

    private fun addBGColorChange(sheetId: Int, range: String, c: Color?): RequestBuilder {
        val split = range.split(":")
        val s1 = split[0]
        val s2 = if (split.size == 1) s1 else split[1]
        return addBatch(
            Request().setUpdateCells(
                UpdateCellsRequest().setRange(buildGridRange(range, sheetId))
                    .setFields("userEnteredFormat.backgroundColor").setRows(
                        getCellsAsRowData(
                            CellData().setUserEnteredFormat(CellFormat().setBackgroundColor(c)),
                            getColumnFromRange(s2) - getColumnFromRange(s1) + 1,
                            getRowFromRange(s2) - getRowFromRange(s1) + 1
                        )
                    )
            )
        )
    }

    fun addCellFormatChange(sheetId: Int, range: String, format: CellFormat, vararg fields: String): RequestBuilder {
        val split = range.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val s1 = split[0]
        val s2 = if (split.size == 1) s1 else split[1]
        return addBatch(
            Request().setUpdateCells(
                UpdateCellsRequest().setRange(buildGridRange(range, sheetId))
                    .setFields("userEnteredFormat(${fields.joinToString(",")})").setRows(
                        getCellsAsRowData(
                            CellData().setUserEnteredFormat(format),
                            getColumnFromRange(s2) - getColumnFromRange(s1) + 1,
                            getRowFromRange(s2) - getRowFromRange(s1) + 1
                        )
                    )
            )
        )
    }

    fun addBGColorChange(sheetId: Int, x: Int, y: Int, c: Color?): RequestBuilder {
        return addBGColorChange(sheetId, x.xc() + y, c)
    }

    fun addNoteChange(sheetId: Int, range: String, note: String?): RequestBuilder {
        val split = range.split(":")
        val s1 = split[0]
        val s2 = if (split.size == 1) s1 else split[1]
        return addBatch(
            Request().setUpdateCells(
                UpdateCellsRequest().setRange(buildGridRange(range, sheetId)).setFields("note").setRows(
                    getCellsAsRowData(
                        CellData().setNote(note),
                        getColumnFromRange(s2) - getColumnFromRange(s1) + 1,
                        getRowFromRange(s2) - getRowFromRange(s1) + 1
                    )
                )
            )
        )
    }

    fun addHorizontalAlignmentChange(sheetId: Int, range: String, alignment: String?): RequestBuilder {
        val split = range.split(":")
        val s1 = split[0]
        val s2 = if (split.size == 1) s1 else split[1]
        return addBatch(
            Request().setUpdateCells(
                UpdateCellsRequest().setRange(buildGridRange(range, sheetId))
                    .setFields("userEnteredFormat.horizontalAlignment").setRows(
                        getCellsAsRowData(
                            CellData().setUserEnteredFormat(CellFormat().setHorizontalAlignment(alignment)),
                            getColumnFromRange(s2) - getColumnFromRange(s1) + 1,
                            getRowFromRange(s2) - getRowFromRange(s1) + 1
                        )
                    )
            )
        )
    }

    fun addVerticalAlignmentChange(sheetId: Int, range: String, alignment: String?): RequestBuilder {
        val split = range.split(":")
        val s1 = split[0]
        val s2 = if (split.size == 1) s1 else split[1]
        return addBatch(
            Request().setUpdateCells(
                UpdateCellsRequest().setRange(buildGridRange(range, sheetId))
                    .setFields("userEnteredFormat.verticalAlignment").setRows(
                        getCellsAsRowData(
                            CellData().setUserEnteredFormat(CellFormat().setVerticalAlignment(alignment)),
                            getColumnFromRange(s2) - getColumnFromRange(s1) + 1,
                            getRowFromRange(s2) - getRowFromRange(s1) + 1
                        )
                    )
            )
        )
    }

    fun addFontChange(sheetId: Int, range: String, font: String?): RequestBuilder {
        val split = range.split(":")
        val s1 = split[0]
        val s2 = if (split.size == 1) s1 else split[1]
        return addBatch(
            Request().setUpdateCells(
                UpdateCellsRequest().setRange(buildGridRange(range, sheetId))
                    .setFields("userEnteredFormat.textFormat.fontFamily").setRows(
                        getCellsAsRowData(
                            CellData().setUserEnteredFormat(CellFormat().setTextFormat(TextFormat().setFontFamily(font))),
                            getColumnFromRange(s2) - getColumnFromRange(s1) + 1,
                            getRowFromRange(s2) - getRowFromRange(s1) + 1
                        )
                    )
            )
        )
    }

    fun addCopyPasteChange(sheetId: Int, range: String, target: String, pasteType: String): RequestBuilder {
        return addBatch(
            Request().setCopyPaste(
                CopyPasteRequest().setSource(buildGridRange(range, sheetId))
                    .setDestination(buildGridRange(target, sheetId)).setPasteType(pasteType)
                    .setPasteOrientation("NORMAL")
            )
        )
    }

    fun addCutPasteChange(sheetId: Int, range: String, target: String, pasteType: String): RequestBuilder {
        return addBatch(
            Request().setCutPaste(
                CutPasteRequest().setSource(buildGridRange(range, sheetId))
                    .setDestination(buildGridCoordinate(target, sheetId)).setPasteType(pasteType)
            )
        )
    }

    class ConditionalFormat(val value: String, val format: CellFormat)

    fun addConditionalFormatCustomFormula(format: ConditionalFormat, range: String, id: Int): RequestBuilder {
        addBatch(
            Request().setAddConditionalFormatRule(
                AddConditionalFormatRuleRequest().setIndex(0).setRule(
                    ConditionalFormatRule().setRanges(listOf(buildGridRange(range, id))).setBooleanRule(
                        BooleanRule().setCondition(
                            BooleanCondition().setType("CUSTOM_FORMULA")
                                .setValues(listOf(ConditionValue().setUserEnteredValue(format.value)))
                        ).setFormat(format.format)
                    )
                )
            )
        )
        return this
    }

    fun addStrikethroughChange(sheetId: Int, range: String, strikethrough: Boolean): RequestBuilder {
        val split = range.split(":")
        val s1 = split[0]
        val s2 = if (split.size == 1) s1 else split[1]
        return addBatch(
            Request().setUpdateCells(
                UpdateCellsRequest().setRange(buildGridRange(range, sheetId))
                    .setFields("userEnteredFormat.textFormat.strikethrough").setRows(
                        getCellsAsRowData(
                            CellData().setUserEnteredFormat(
                                CellFormat().setTextFormat(
                                    TextFormat().setStrikethrough(
                                        strikethrough
                                    )
                                )
                            ),
                            getColumnFromRange(s2) - getColumnFromRange(s1) + 1,
                            getRowFromRange(s2) - getRowFromRange(s1) + 1
                        )
                    )
            )
        )
    }

    fun addStrikethroughChange(sheetId: Int, x: Int, y: Int, strikethrough: Boolean): RequestBuilder {
        return addStrikethroughChange(sheetId, x.xc() + y, strikethrough)
    }

    fun addFGColorChange(sheetId: Int, range: String, c: Color?): RequestBuilder {
        val split = range.split(":")
        val s1 = split[0]
        val s2 = if (split.size == 1) s1 else split[1]
        return addBatch(
            Request().setUpdateCells(
                UpdateCellsRequest().setRange(buildGridRange(range, sheetId))
                    .setFields("userEnteredFormat.textFormat.foregroundColor").setRows(
                        getCellsAsRowData(
                            CellData().setUserEnteredFormat(CellFormat().setTextFormat(TextFormat().setForegroundColor(c))),
                            getColumnFromRange(s2) - getColumnFromRange(s1) + 1,
                            getRowFromRange(s2) - getRowFromRange(s1) + 1
                        )
                    )
            )
        )
    }

    fun addFGColorChange(sheetId: Int, x: Int, y: Int, c: Color?): RequestBuilder {
        return addFGColorChange(sheetId, x.xc() + y, c)
    }

    private val userEntered: List<ValueRange>
        get() = requests.filter { it.valueInputOption == ValueInputOption.USER_ENTERED }.map { it.build() }
    private val raw: List<ValueRange>
        get() = requests.filter { it.valueInputOption == ValueInputOption.RAW }.map { it.build() }
    private val batch: List<Request>
        get() = requests.filter { it.valueInputOption == ValueInputOption.BATCH }.map { it.buildBatch() }

    fun clear() {
        requests.clear()
        executed = false
    }

    fun execute(realExecute: Boolean = true) {
        this.realExecute = realExecute
        scheduleForExecution(this)
    }

    /**
     * Executes the request to the specified google sheet
     */
    private fun internalExecute(tries: Int = 0) {
        check(!executed) {
            """
     Already executed RequestBuilder with requests:
     sid = $sid
     $requests
     """.trimIndent()
        }
        executed = true
        val userentered = userEntered
        if (!realExecute) {
            printUserEntered(userentered)
            return
        }
        val raw = raw
        val batch = batch
        val job = scope.async {
            launch {
                if (batch.isNotEmpty()) {
                    Google.batchUpdate(sid, batch)
                    additionalSheets?.forEach {
                        Google.batchUpdate(it, batch)
                    }
                }
            }
            if (!onlyBatch) {
                launch {
                    if (userentered.isNotEmpty()) {
                        if (!suppressMessages) printUserEntered()
                        Google.batchUpdate(sid, userentered, "USER_ENTERED")
                        additionalSheets?.forEach {
                            Google.batchUpdate(it, userentered, "USER_ENTERED")
                        }
                    }
                }
                launch {
                    if (raw.isNotEmpty()) {
                        Google.batchUpdate(sid, raw, "RAW")
                        additionalSheets?.forEach {
                            Google.batchUpdate(it, raw, "RAW")
                        }
                    }
                }
            }
        }
        scope.launch {
            try {
                job.await()
            } catch (ex: GoogleJsonResponseException) {
                if (ex.details.code == 401) {
                    val newTries = tries + 1
                    val timeMillis = newTries * 3000L
                    logger.info("Got 401, trying again in {}ms", timeMillis)
                    delay(timeMillis)
                    if (newTries < 3) {
                        internalExecute(newTries)
                    } else {
                        throw ex
                    }
                } else {
                    throw ex
                }
            }
        }
        runnable?.let {
            scope.launch {
                job.join()
                delay(delay)
                it()
            }
        }
    }

    fun executeAndReset(realExecute: Boolean = true) {
        execute(realExecute)
        clear()
    }

    fun printUserEntered(userentered: List<ValueRange> = userEntered) {
        logger.info("RequestBuilder with requests:")
        logger.info("sid = $sid")
        for (i in userentered.indices) {
            val range = userentered[i]
            logger.info("{}: {} -> {}", i, range.range, range.values)
        }
    }

    private enum class ValueInputOption {
        RAW, USER_ENTERED, BATCH
    }

    private class MyRequest {
        private var range: String? = null
        private var send: List<List<Any>?>? = null
        var valueInputOption: ValueInputOption? = null
        private var request: Request? = null
        fun setRange(range: String?): MyRequest {
            this.range = range
            return this
        }

        fun setSend(send: List<List<Any>?>?): MyRequest {
            this.send = send
            return this
        }

        fun setRequest(request: Request?): MyRequest {
            this.request = request
            valueInputOption = ValueInputOption.BATCH
            return this
        }

        fun setValueInputOption(valueInputOption: ValueInputOption?): MyRequest {
            this.valueInputOption = valueInputOption
            return this
        }

        fun build(): ValueRange {
            return ValueRange().setValues(send).setRange(range)
        }

        fun buildBatch(): Request {
            return request!!
        }

        override fun toString(): String {
            return "Request{range='$range', send=$send, valueInputOption='$valueInputOption'}"
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(RequestBuilder::class.java)
        private val EVERYTHING_BUT_NUMBER = Pattern.compile("\\D")
        private val EVERYTHING_BUT_CHARS = Pattern.compile("[^a-zA-Z]")
        private val scope = createCoroutineScope("RequestBuilder")
        private val channelCollectScope = createCoroutineScope("RequestBuilderChannelCollect")
        private val channel = Channel<RequestBuilder>(Channel.UNLIMITED)

        init {
            channelCollectScope.launch {
                while (true) {
                    channel.receive().internalExecute()
                    delay(2000)
                }
            }
        }

        private fun scheduleForExecution(b: RequestBuilder) {
            defaultScope.launch(start = CoroutineStart.UNDISPATCHED) { channel.send(b) }
        }

        fun getColumnFromRange(range: String): Int {
            val chars = EVERYTHING_BUT_CHARS.matcher(range).replaceAll("").toCharArray()
            return if (chars.size == 1) chars[0].code - 65 else (chars[0].code - 64) * 26 + (chars[1].code - 65)
        }

        fun getRowFromRange(range: String): Int {
            return EVERYTHING_BUT_NUMBER.matcher(range).replaceAll("").toInt() - 1
        }

        fun buildGridRange(expr: String, sheetId: Int): GridRange {
            val split = expr.split(":")
            val s1 = split[0]
            val s2 = if (split.size == 1) s1 else split[1]
            val r = GridRange()
            r.sheetId = sheetId
            r.setStartColumnIndex(getColumnFromRange(s1)).startRowIndex = getRowFromRange(s1)
            r.setEndColumnIndex(getColumnFromRange(s2) + 1).endRowIndex = getRowFromRange(s2) + 1
            logger.info("r = {}", r.toPrettyString())
            return r
        }

        fun buildGridCoordinate(expr: String, sheetId: Int): GridCoordinate {
            val split = expr.split(":")
            val s1 = split[0]
            val r = GridCoordinate().apply {
                this.sheetId = sheetId
                columnIndex = getColumnFromRange(s1)
                rowIndex = getRowFromRange(s1)
            }
            logger.info("r = {}", r.toPrettyString())
            return r
        }
    }
}
